topics/basics/testing_plugins/integration_tests/integration_tests_ui.md

```diff
@@ -0,0 +1,198 @@
+
+So, being explicit about the component hierarchy helps create more robust and maintainable UI automation code.
+
+## Searching Components
+
+While the Driver framework provides many pre-built components (like `ideFrame`, `codeEditor`, `button`, `tree`, etc.), sometimes it's required to locate custom elements.
+
+One way to pause the IDE to examine its UI structure:
+
+```kotlin
+.runIdeWithDriver().useDriverAndCloseIde {
+  Thread.sleep(30.minutes.inWholeMilliseconds)
+}
+```
+
+When the test is running, the following line will appear in the logs: `http://localhost:63343/api/remote-driver/`.
+Opening this URL reveals an HTML representation of the IDE's Swing component tree:
+
+![](integration_tests_devtools.png){width=706}
+
+Using Developer Tools in the browser, it's possible to inspect detailed component attributes.
+Here's an example component:
+
+```html
+
+<div accessiblename="Current File"
+     asstring="<result of calling toString on component>"
+     class="ActionButtonWithText"
+     enabled="true" hide_dropdown_icon="HIDE_DROPDOWN_ICON"
+     javaclass="com.intellij.execution.ui.RedesignedRunConfigurationSelector$createCustomComponent$1"
+     myaction="Select Run/Debug Configuration (null)"
+     tool_tip_text_key="ToolTipText"
+     visible="true" visible_text="Current File" visible_text_keys=""/>
+```
+
+There are other attributes which are omitted for clarity.
+
+The element corresponds to the following button:
+
+![](integration_tests_ui_sample.png){width=706}
+
+Similar to web testing frameworks like Selenium, XPath is used to locate components.
+The Driver framework provides a simple XPath builder.
+Here are several ways to find the same component:
+
+```kotlin
+xQuery { byVisibleText("Current File") }
+xQuery { byAccessibleName("Current File") }
+xQuery { byType("com.intellij.execution.ui.RedesignedRunConfigurationSelector\$createCustomComponent$1") }
+```
+
+For reliable component identification, prioritize these attributes: `accessiblename`, `visible_text`, `icon`, `javaclass`.
+Multiple attributes can be combined for a more precise selection.
+
+## Interaction with Components
+
+Once a component is located, it's possible to interact with it or verify its properties.
+
+To click the <control>Current File</control> button:
+
+```kotlin
+x(xQuery { byVisibleText("Current File") }).click()
+```
+
+The `x()` call creates a lazy reference to the component.
+It means that the XPath query isn't executed immediately and component lookup happens only when an action (like `click()`) is invoked.
+
+Here's a part of test that incorporates UI interaction:
+```kotlin
+runIdeWithDriver().useDriverAndCloseIde {
+  waitForIndicators(1.minutes)
+  ideFrame {
+    x(xQuery { byVisibleText("Current File") }).click()
+  }
+}
+```
+
+Beyond mouse clicks, keyboard input and shortcuts can be simulated:
+
+```kotlin
+keyboard {
```

Comment:
Can we extend this example so readers can understand how to use `keyboard` and type in a component? It is unclear to me how to do it. Is there a component focus request needed first?

Reply:
I've added a general information about the focus and the recommended way to do this.